mgr/src/server/plugins/volume_heal.cr

```diff
@@ -0,0 +1,380 @@
+  end
+
+  # The volume must be mounted else, heal is not possible.
+  client_volfile = "/var/lib/kadalu/volfiles/client-dev-#{volume_name}.vol"
+
+  if !File.exists?(client_volfile)
+    halt(env, status_code: 400, response: ({"error": "The client volfile does not exist for volume(#{volume_name}). Mount and retry!"}.to_json))
+  end
+
+  puts "IN HEAL INFO"
+
+  rc, output, err = execute("/glusterfs/heal/src/glfsheal", ["vol1", "info-summary", "--xml", "volfile-path", client_volfile])
+  puts "rcs: #{rc}, output: #{output}, error: #{err}"
+
+  if rc < 0
+    halt(env, status_code: 400, response: ({"error": err}.to_json))
+  end
+
+  document = XML.parse(output)
+
+  p = document.first_element_child
+  if p
+    puts "P: #{p}"
+  end
+
+  bricks = document.xpath_nodes("//bricks")
+  puts "document: #{document}, storage_unit:, bricks: #{bricks}"
+
+  brick_data = Hash(String, HealStorageUnit).new
+
+  bricks.map do |bk|
+    # puts "bk_id: #{bk["hostUuid"]}"
+    brick = HealStorageUnit.new
+    bk.children.each do |ele|
+
+      # puts "ele.name: #{ele.name}"
+      next unless ele.name == "brick"
+      puts "ele_hostUuid: #{ele["hostUuid"]}"
+      puts "ele.name: #{ele.name}"
+      puts "ele.content: #{ele.content}"
+
+      next unless ele["hostUuid"] != "-"
+
+      brick.id = ele["hostUuid"].strip
+      ele.children.each do |ele_child|
+        puts "ele_child.name :#{ele_child.name}"
+        case ele_child.name
+        # when "hostUuid"
+        #   brick.id = ele_child.content.strip
+        #   puts brick.id
+        when "name"
+          brick.name = ele_child.content.strip
+          # puts ele_child.content.strip
+        when "status"
+          brick.status = ele_child.content.strip
+
+        # when "file"
+        #   # brick.heal_pending_files << {ele_child["gfid"], ele_child.content.strip}
+        #   puts "ele_child gfid: #{ele_child["gfid"]}"
+
+        when "totalNumberOfEntries"
+          brick.totalNumberOfEntries = ele_child.content.strip
+        when "numberOfEntriesInHealPending"
+          brick.numberOfEntriesInHealPending = ele_child.content.strip
+        when "numberOfEntriesInSplitBrain"
+          brick.numberOfEntriesInSplitBrain = ele_child.content.strip
+        when "numberOfEntriesPossiblyHealing"
+          brick.numberOfEntriesPossiblyHealing = ele_child.content.strip
+        end
+      end
+      puts "brick: #{brick}"
+      brick_data[brick.name] = brick
+    end
+    puts "br: #{brick_data.to_json}"
+  end
+
+  volume.distribute_groups.each do |dist_grp|
+    dist_grp.storage_units.each do |storage_unit|
+      puts "storage_unit name: #{storage_unit.node.name}, #{storage_unit.path}"
+
+      storage_unit_full_path = storage_unit.node.name + ":" + storage_unit.path
+
+      if brick_data.has_key?(storage_unit_full_path)
+        heal_storage_unit_data = brick_data[storage_unit_full_path]
```

Comment:
```crystal
heal_storage_unit_data = brick_data[storage_unit_full_path]?
unless heal_storage_unit_data.nil?
```

Reply:
isnt it,
`next unless !heal_storage_unit_data.nil?` ??